src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/HttpRequestExtensions.cs

```diff
@@ -62,4 +67,34 @@ public static bool Accepts(this HttpRequest request, bool exact = false, params
             mediaTypes.Any(type => exact ? type.Equals(acceptedType) : type.IsSubsetOf(acceptedType))
         ) ?? false;
     }
+
+    public static async Task<TJsonType?> GetJsonBody<TJsonType>(this HttpRequest request, bool allowEmptyBody = true)
+        where TJsonType : class
+    {
+        var requestBody = await new StreamReader(request.Body).ReadToEndAsync();
+
+        if (requestBody.IsNullOrEmpty() && allowEmptyBody)
+        {
+            return null;
+        }
+
+        return JsonConvert.DeserializeObject<TJsonType>(requestBody) ??
+               throw new ArgumentException($"Could not deserialize request body to type {typeof(TJsonType)}");
+    }
+
+    public static bool TryGetHeader(
+        this HttpRequest httpRequest,
+        string headerName,
+        [MaybeNullWhen(false)] out string headerValue)
```

Comment:
Good point, thanks for the article. I've made this change now, although struggling to get rid of the null-forgiving operator in the usages of this method now.

I added the check `previewTokenValue.Count == 1` to ensure `previewTokenValue` is a `string` rather than `null` but don't think it's super necessary.

```csharp
        return httpContextAccessor.HttpContext.TryGetRequestHeader(
                   RequestHeaderNames.PreviewToken,
                   out var previewTokenValue)
               && previewTokenValue.Count == 1
               && await previewTokenService.ValidatePreviewToken(previewTokenValue!, dataSetVersion.Id);
```